hist/histv7/inc/ROOT/RAxis.hxx

```diff
@@ -96,6 +96,67 @@ protected:
       return (int)rawbin;
    }
 
+   /// Check if two axis have the same bin borders
+   ///
+   /// Default implementation should work for any RAxis type, but is a little
+   /// bit stupid. RAxis implementations are encouraged to provide optimized
+   /// overrides for common comparison scenarios.
+   virtual bool HasSameBinBordersAs(const RAxisBase& other) const noexcept {
+      // Axis growability (and thus under/overflow bin existence) must match
+      if (CanGrow() != other.CanGrow()) return false;
+
+      // Number of normal bins must match
+      if (GetNBinsNoOver() != other.GetNBinsNoOver()) return false;
+
+      // Left borders of normal bins must match
+      for (int bin: *this) {
+         if (GetBinFrom(bin) != other.GetBinFrom(bin)) return false;
+      }
+
+      // Right border of the last normal bin (aka maximum) must also match
+      if (GetMaximum() != other.GetMaximum()) return false;
+
+      // If all of these checks passed, the two axes have the same bin borders
+      return true;
+   }
+
+   /// If supplementary bin metadata (e.g. bin labels) exists in `this`, make
+   /// sure that `other` provides the same bin metadata.
+   ///
+   /// The implementation can work under the assumption that the number of bins
+   /// and bin borders of `this` and `other` has already been compared equal.
+   ///
+   /// RAxis classes which override a non-default implementation of
+   /// `HasSameBinMetadataAs` should start by calling the parent class'
+   /// implementation of this method, in order to allow "stacking" of bin
+   /// metadata across the class hierarchy.
+   ///
+   /// Adding an override of `HasSameBinMetadataAs` to an RAxis class is
+   /// considered an API-breaking change: subclasses can assume that if their
+   /// parent classes don't override the default implementation of
+   /// `HasSameBinMetadataAs`, they never will.
+   virtual bool HasSameBinMetadataAs(const RAxisBase& /*other*/) const noexcept {
+      return true;
+   }
+
+   /// Semantically equivalent to `HasSameBinMetadataAs(other)`, except that the
+   /// implementation can be optimized under knowledge of the fact that
+   /// `other.HasSameBinMetadataAs(*this)` has previously been called and
+   /// returned a positive result.
+   ///
+   /// RAxis classes which override a non-default implementation of
+   /// `AlsoHasSameBinMetadataAs` should start by calling the parent class'
+   /// implementation of this method, in order to allow "stacking" of bin
+   /// metadata across the class hierarchy.
+   ///
+   /// Adding an override of `AlsoHasSameBinMetadataAs` to an RAxis class is
+   /// considered an API-breaking change: subclasses can assume that if their
+   /// parent classes don't override the default implementation of
+   /// `AlsoHasSameBinMetadataAs`, they never will.
+   virtual bool AlsoHasSameBinMetadataAs(const RAxisBase& other) const noexcept {
```

Comment:
I agree that this is ugly and welcome alternate design and naming solutions. Let me explain what this ugliness is about and why it is the best I could come up with.

---

Fundamentally, what I'm trying to do here is to emulate, at the library level, a nice programming language feature from the Lisp family which C++ doesn't have. That language feature goes by the names of [multiple dispatch](https://en.wikipedia.org/wiki/Multiple_dispatch) and "multimethods".

Using this language feature, one could define the following overload set...

- `HasSameBinMetadata(RAxisLabels&, RAxisLabels&) = /* code that checks labels */`
    * Overload that gets resolved when both axes have labels, even if they are subclasses of RAxisLabels.
- `HasSameBinMetadata(RAxisLabels&, RAxisBase&) = false`
    * Overload that gets resolved when fist axis has labels and second axis doesn't, if they both had labels the (Labels, Labels) overload would be preferred.
- `HasSameBinMetadata(RAxisBase&, RAxisLabels&) = false`
    * Overload that gets resolved when second axis has labels and first axis doesn't, if they both had labels the (Labels, Labels) overload would be preferred.
- `HasSameBinMetadata(RAxisBase&, RAxisBase&) = true`
    * Overload that gets resolved neither axis has labels, otherwise above overloads would be picked up.

...then call `HasSameBinMetadata` with a pair of `RAxisLabels&` reference-to-base, and have that call dispatch _at runtime_ to the right version of `HasSameBinMetadata` depending on the actual pair of axis types at hand.

There are several well-established tricks to emulate multiple dispatch in languages like C++ which only have single dispatch...

- Visitor pattern
- Single `HasSameBinMetadata(RAxisLabels&, RAxisBase&)` implementation with a big bunch of dynamic_casts and if statements.
- Global hashmap from a pair of RAxis typeid to concrete `HasSameBinMetadata` implementation.

...however, all of these tricks crucially rely on having exhaustive knowledge of all classes in the `RAxis` class hierarchy. This would go against our desire to keep said class hierachy open (not to mention that those implementations rapidly get uglier as the amount of RAxis types increases, so even implementing every possible RAxis type inside of ROOT wouldn't be a super-nice solution).

Which is why I came up with what I believe to be a new approach to emulating multimethods in C++, that happens to work well for the `HasSameBinMetadata` use case.

---

Fundamentally, the reason why we can't have just `virtual bool HasSameBinMetadataAs(const RAxisBase& other)` is that in this single-dispatch design, we cannot make `axis_eq.HasSameBinMetadataAs(axis_labels)` work without making `RAxisEquidistant` aware of the existence of `RAxisLabels` (which would break the extensibility of the `RAxis` class hierarchy.

An obvious fix is to test both `this->HasSameBinMetadataAs(other)` and `other.HasSameBinMetadataAs(*this)`, which will work no matter if `this` or `other` has metadata that the other isn't aware of. However, this approach leads to redundant computation when `this` and `other` are of the same type (e.g. duplicate labels comparisons).

This is the motivation between the optional `AlsoHasSameBinMetadataAs` extension point, which allows the second metadata comparison to optimize under knowledge of the fact that the first metadata comparison succeeded, in this case allowing deduplication of labels comparions.

---

As far as I can tell, this is one of the simplest designs under the following constraints:

1. RAxis class hierarchy is user-extensible (and therefore, implementation cannot assume omniscient knowledge of the full `RAxisBase` class hierarchy)
2. RAxisBase subclasses can add metadata which RAxisBase itself doesn't know about, such as axis labels.
3. Metadata mismatch should be an error in the eye of `HasSameBinningAs` comparisons.
4. Axis metadata comparisons should not do duplicate work.

But if you can think of a better way to do it, or if you think that some of the above design constraints aren't worthwhile and can be relaxed, please tell me.

Reply:
To summarize the RHist meeting discussion, we're going to resolve this issue by removing constraint number 2: RAxisBase will know about labels, and we won't allow user to add new bin metadata without us knowing about it.

Reply:
Implemented in d3ac8bf85bdbe .